Produce basic JSON summary of a specified package

CicadaCinema opened the pull request:
This will include the names of public symbols and their kind (function, variable, class, etc)

CicadaCinema committed:
specify package location and run the analyser on every dart file in my_package/lib

CicadaCinema commented on lib/src/package_analysis.dart:
```diff
@@ -0,0 +1,60 @@
+import 'package:path/path.dart' as path;
+
+Future<void> main(List<String> arguments) async {
+  // parse command-line arguments
```
The analyser will have a number of distinct functions (creating a JSON summary, comparing two JSON summaries, comparing two package versions directly), so it probably makes sense to introduce subcommands for each of these functions

sigurdm replied:
Yes, but also the functionality can be exposed as dart api, such that pana can call directly into it.

CicadaCinema replied:
okay - for now I think I will try to create a separate CLI interface which will call the dart api (at least that's what I think I am doing)

CicadaCinema commented on lib/src/package_analysis.dart:
```diff
@@ -0,0 +1,60 @@
+    final session = context.currentSession;
+    final someResolvedLibrary = await session.getResolvedLibrary(filePath);
+
+    // match [packageLocation]/lib/*.dart
```
There are 3 kinds of libraries in a package which can be imported:
1. `lib/my_package.dart` (we usually import this library when we want to use the package)
2. `lib/*.dart` (could be in a sub-folder of `lib`, eg `lib/feature/*.dart`)
3. `lib/src/*.dart` (not in a sub-folder)

Even type 3 libraries can be imported for use, but it might make sense to exclude these libraries from analysis (by default, adding a switch to disable this).

sigurdm replied:
Yeah - anything under lib/src (also sub-folders) is considered private to the package by convention, and should not be included in the summary of the exported API surface.

https://dart.dev/guides/libraries/create-library-packages#what-makes-a-library-package

CicadaCinema committed:
implement CLI interface with CommandRunner
(using sample in https://pub.dev/packages/args#dispatching-commands )
also exclude [packageLocation]/lib/src/*.dart when searching for libraries

jonasfj commented on lib/src/package_analysis/summary.dart:
```diff
@@ -0,0 +1,49 @@
+Future<void> generateSummary(String packageLocation) async {
+  if (!await FileSystemEntity.isDirectory(packageLocation)) {
+    exitFatal('Please specify a directory for analysis.', 1);
+  }
```
I would suggest throwing a custom subclass of `Exception`. Then catching that in `SummaryCommand` and exiting non-zero.

jonasfj commented on lib/src/package_analysis/summary.dart:
```diff
@@ -0,0 +1,49 @@
+import './common.dart';
+
+Future<void> generateSummary(String packageLocation) async {
+  if (!await FileSystemEntity.isDirectory(packageLocation)) {
```
```suggestion
  if (!await Directory.exists(packageLocation)) {
```

jonasfj replied:
I like this sanity check, we might want to check that `$packageLocation/.dart_tool/package_config.json` exists.

In short, `dart pub get` will:
 * Solve dependencies to satisfy constraints (only one version of each package).
 * Download/extract dependencies into `$PUB_CACHE` (default `$HOME/.pub-cache/`)
 * Write a `pubspec.lock` file (if one doesn't exist)
 * Write a `.dart_tool/package_config.json` file which maps package names to folders in `PUB_CACHE`.
  (This is how dependencies are resolved).

Details see:
https://github.com/dart-lang/language/blob/master/accepted/2.8/language-versioning/package-config-file-v2.md

This isn't important, but I'm guessing we don't want to analyze code without fetching dependencies first. Ofcourse it's possible that we only want to do a AST level analysis. There are pros/cons here.. If we do an AST level analysis we won't be able to see which methods are inherited when creating a subclass of a class from a dependency, or at-least if we want to we'll have to find the summary of the dependency and do some work...

CicadaCinema replied:
Yeah, that sounds like a good idea. Is there a way for the analyser (or some other API) to fetch dependencies? I'm guessing not, we'll have to throw an exception and ask the user to run `dart pub get` themselves (if running through the CLI), or use `Process.run()`. Neither option is ideal really...

CicadaCinema replied:
added an check for `File(path.join(packageLocation, '.dart_tool', 'package_config.json')).exists()`

jonasfj commented on lib/src/package_analysis/summary.dart:
```diff
@@ -0,0 +1,49 @@
+      // use canonicalize to ensure we get an absolute path
+      includedPaths: [path.canonicalize(packageLocation)],
+    ).first,
+  );
```
Consider using: `AnalysisContextCollection`, see:
https://pub.dev/documentation/analyzer/latest/dart_analysis_analysis_context_collection/AnalysisContextCollection-class.html

jonasfj replied:
Be ware that packages may contain an `analysis_options.yaml` file.
We can use a `OverlayResourceProvider` to override this file in-memory, if we want to:
https://github.com/dart-lang/pub/blob/ea070238f14d31e28fe448cb136d31bed96705e5/lib/src/dart.dart#L45-L53

It has pros/cons, if we override it then users can't opt files out of analysis. On the other hand, we might not want to analyse the package with whatever options they've enabled.

jonasfj commented on lib/src/package_analysis/summary.dart:
```diff
@@ -0,0 +1,49 @@
+
+    // ensure that resolving has been successful
+    if (someResolvedLibrary is! ResolvedLibraryResult) {
+      stderr.writeln('$filePath is not a library');
```
Isn't it more like that analysis of the specified file failed?

jonasfj replied:
Also instead of writing to `stderr` consider creating an context object like:
```dart
abstract class PackageAnalysisContext {
  AnalysisContextCollection get analysisContextCollection;
  AnalysisContext contextFor(String path) => analysisContextCollection.contextFor(path);

  /// Log [message] as warning that something unexpected happened.
  void warning(String message);
}
```

Then people calling `generateSummary` can provide an instance of `PackageAnalysisContext` and they can decide whether to print to stderr or write a log file.

CicadaCinema replied:
> Isn't it more like that analysis of the specified file failed?

More specifically, I might say 'Building a resolved AST for this library returned an invalid result.' (Not sure if this is too verbose)



> Then people calling `generateSummary` can provide an instance of `PackageAnalysisContext` and they can decide whether to print to stderr or write a log file.

Right, so they would have to provide their own handler? In case of a fatal error I could throw an exception instead of calling the handler (well, I suppose I could call the handler, let the user log the error and THEN throw the exception), like `InconsistentAnalysisException` in the case of `AnalysisSession`.

jonasfj commented on lib/src/package_analysis/summary.dart:
```diff
@@ -0,0 +1,49 @@
+
+void _traverseLibrary(LibraryElement libraryElement) {
+  print(libraryElement.identifier);
+  // TODO
```
Take a look at, various visitors:
https://pub.dev/documentation/analyzer/latest/dart_element_element/ElementVisitor-class.html

Also it's probably a good idea if we read the tutorial:
https://github.com/dart-lang/sdk/blob/main/pkg/analyzer/doc/tutorial/tutorial.md

It's probably a little outdated/incomplete, like I think new types have been introduced with the null-safety migration.

CicadaCinema replied:
hmm, I've read those but I'm still not quite sure how visitors would be useful in this case (compared to getters), because:
- we're not interested in private elements such as functions inside functions
- we don't have a goal as broad as 'count the number of elements of a certain kind in an entire CompilationUnit'

Also, the schema of the JSON summary I have in mind is rather rigid - I would assume it 'mirrors' the way getters are used to traverse the AST structure:

<details>
  <summary>what the JSON summary could look like</summary>

```json
{
  "libraries": [
    {
      "name": "library_1",
      "exports": [
        {
          "name": "myClass",
          "kind": "class"
        },
        {
          "name": "myFunction",
          "kind": "function"
        }
      ]
    }
  ]
}
```
</details>

Still, I concede that the visitor model is probably more flexible and will pay off in the long run. Given that we're still two weeks away from the start of the coding period, maybe I will experiment with visitors, instead of repeating the getter model I used in my sample project.

CicadaCinema replied:
Also reading that document, I realise I haven't touched a `CompilationUnit` at all in my sample project at all - I obtained a `LibraryElement` instead for every .dart file in the package folder. What's the difference between the two?

It looks like both represent a resolved AST, but a library has to be explicitly defined using the `library` directive (and can be split between different files using `part`). Does every .dart file have its own `CompilationUnit` and vice versa?

It looks like symbols don't have to be part of a library in order to be public and importable, so I should probably work with units instead of libraries. Right?

sigurdm replied:
I think each dart file is represented by a `CompilationUnit`, and a LibraryElement that is split in parts is then represented as several `CompilationUnit`s

(I think `CompilationUnit` is kind of a mis-nomer as I would expect the unit of compilation to actually be the library...)

jonasfj reviewed (commented):
Great start!

CicadaCinema committed:
add test for the `part` directive
(the current version of the analyser deals with this type of scenario just fine)

CicadaCinema committed:
deal with the case where we encounter a file which is just part of another library
also test if `packageLocation` exists at all

CicadaCinema committed:
begin refactoring the implementation of warnings/errors so they are not confined to just the CLI case scenario

jonasfj commented on lib/src/package_analysis.dart:
```diff
@@ -23,10 +26,24 @@ class SummaryCommand extends Command {
 
     if (paths.length != 1) {
-      exitFatal('Please specify exactly one directory for analysis.', 1);
+      throw const PackageAnalysisError(
```
You can also use `ArgumentError`

CicadaCinema committed:
delegate argument handling to CLI interface library
also create analysis context collection there

CicadaCinema committed:
create shape classes,
add basic logic to analyse libraries to produce a PackageShape

CicadaCinema committed:
use warning method instead of printing to stderr

CicadaCinema committed:
make shapes serialisable and print the results of summarising the given package

CicadaCinema committed:
ensure private methods are not included in ClassShape

CicadaCinema committed:
enable the summary to show exported libraries
also write tests for various features of the analyser

CicadaCinema committed:
sort filenames being analysed

jonasfj commented on test/package_analysis.dart:
```diff
@@ -0,0 +1,81 @@
+Future<void> main() async {
+  final yamlDir =
+      Directory(path.join(path.current, 'test', 'package_analysis'));
+  final configTemplate = '''{
+  "configVersion": 2,
+  "packages": [
+    {
+      "name": "test_package",
+      "rootUri": "../",
+      "packageUri": "lib/",
+      "languageVersion": "2.10"
+    }
+  ],
+  "generated": "2022-01-01T00:00:00.000000Z",
+  "generator": "pub",
+  "generatorVersion": "2.17.3"
+}
+''';
```
Consider:
```suggestion
  final configTemplate = json.encode({
  "configVersion": 2,
  "packages": [
    {
      "name": "test_package",
      "rootUri": "../",
      "packageUri": "lib/",
      "languageVersion": "2.10"
    }
  ],
  "generated": "2022-01-01T00:00:00.000000Z",
  "generator": "pub",
  "generatorVersion": "2.17.3"
});
```

jonasfj replied:
Also consider not using language version 2.10 anywhere... ensure that lower-bound SDK constraint is always `>= 2.12`, so that you've opted into null-safety.

CicadaCinema replied:
alright! not sure why languageVersion isn't the same as generatorVersion here, though - I've had Dart 2.17.3 installed on my machine throughout

jonasfj replied:
See: https://dart.dev/guides/language/evolution#language-versioning

jonasfj commented on test/package_analysis/basic.yaml:
```diff
@@ -0,0 +1,37 @@
+---
```
nit: there is no specification for it in dart... but in golang, people tend to through data used for testing into a testdata/ folder.

Consider a folder layout like:
```
lib/
  src/
    package_analysis/
      package_analysis.dart // exports everything you want to make publicly visible
      summary.dart
      ... // all the other files
test/
  package_analysis/
    package_analysis_test.dart
    testdata/
      basic.yaml
      ...
```

jonasfj commented on lib/src/package_analysis/shape.dart:
```diff
@@ -0,0 +1,56 @@
+@JsonSerializable()
+class LibraryShape {
+  final String identifier;
+  @JsonKey(name: 'exported-classes')
```
Prefer `exportedClasses`, see: https://google.github.io/styleguide/jsoncstyleguide.xml

jonasfj commented on lib/src/package_analysis/shape.dart:
```diff
@@ -0,0 +1,56 @@
+
+@JsonSerializable()
+class LibraryShape {
+  final String identifier;
```
Should we call this `path`, `importPath` or `filepath` or? maybe `path` is fine?

CicadaCinema replied:
can `package:my_package/library.dart` really be called a path? I specifically didn't call this `path` because it's NOT the filesystem path for this library file

jonasfj replied:
uri?

jonasfj commented on lib/src/package_analysis/shape.dart:
```diff
@@ -0,0 +1,56 @@
+  final String name;
+  final List<MethodShape> methods;
+
+  ClassShape(this.id, this.name, this.methods);
```
You might want to do something like:
```suggestion
  ClassShape({required this.id, required this.name, required this.methods});
```

That way there is less risk something will blow up.

jonasfj commented on lib/src/package_analysis/shape.dart:
```diff
@@ -0,0 +1,56 @@
+
+@JsonSerializable()
+class ClassShape {
+  final int id;
```
This property is probably worth documenting.
```suggestion
  /// Unique identifier for this [ClassShape] within the given [PackageShape].
  ///
  /// Classes will be referenced using their [id] in other shapes.
  final int id;
```

jonasfj commented on lib/src/package_analysis/shape.dart:
```diff
@@ -0,0 +1,56 @@
+class LibraryShape {
+  final String identifier;
+  @JsonKey(name: 'exported-classes')
+  final Set<int> exportedClasses;
```
I would suggest documenting this property too... something saying that this is `[ClassShape.id]` of the classes that are exported.

jonasfj commented on lib/src/package_analysis/summary.dart:
```diff
@@ -0,0 +1,128 @@
+
+import 'common.dart';
+
+class Summarizer {
```
nit: but instead of making a class that a single member that can be called once.

Have you considered creating a function with nested functions?

Because this isn't really a class that you create an instance of and then you do things to it... or pass it around.
It's a class you instantiate and then you call a function on it.

```dart
Future<PackageShape> summarizePackage(
    PackageAnalysisContext packageAnalysisContext,
    String packageLocation,
) async {
  var _classCounter = 0;
  final _package = PackageShape(<LibraryShape>[], <ClassShape>[]);

  /// Ids of classes defined in a given library
  final _classDefinitions = <String, Set<int>>{};

  /// Identifiers of libraries exported in a given library
  final _libraryExports = <String, List<String>>{};

  /// Ids of classes exported in a given library
  final _classExports = <String, Set<int>>{};
  
  //#region Methods for summarizing elements

void summarizeLibraryElement(
      LibraryElement libraryElement, String libraryPath) {
    var identifier = libraryElement.identifier;

    // classes defined in this library
    var classes = libraryElement.topLevelElements
        .where((element) => (element.isPublic && element is ClassElement))
        .map((element) => summarizeClassElement(element as ClassElement))
        .toList();
    var classIds = classes.map((thisClass) => thisClass.id).toSet();

    _package.classes.addAll(classes);
    _package.libraries.add(LibraryShape(identifier, <int>{}));

    _classDefinitions[identifier] = classIds;
    _libraryExports[identifier] = libraryElement.exportedLibraries
        .map((library) => library.identifier)
        .toList();
  }

  ClassShape summarizeClassElement(ClassElement classElement) {
    var methods = classElement.methods
        .where((element) => element.isPublic)
        .map(summarizeMethodElement)
        .toList();
    _classCounter += 1;
    return ClassShape(_classCounter, classElement.name, methods);
  }

  MethodShape summarizeMethodElement(MethodElement methodElement) {
    return MethodShape(methodElement.name);
  }

  //#endregion
  
  var collection = packageAnalysisContext.analysisContextCollection;

    for (var context in collection.contexts) {
      for (var filePath in context.contextRoot.analyzedFiles().sorted()) {
        final session = context.currentSession;
        final library = await session.getResolvedLibrary(filePath);

        // match [packageLocation]/lib/*.dart
        // but exclude [packageLocation]/lib/src/*.dart
        if (!(path.isWithin(path.join(packageLocation, 'lib'), filePath) &&
            !path.isWithin(
                path.join(packageLocation, 'lib', 'src'), filePath) &&
            path.extension(filePath) == '.dart')) {
          continue;
        }

        // this file is just part of another library
        if (library is NotLibraryButPartResult) {
          continue;
        }

        // ensure that resolving has been successful
        if (library is! ResolvedLibraryResult) {
          packageAnalysisContext
              .warning('analysis of $filePath as a library failed');
          continue;
        }

        summarizeLibraryElement(
            library.element, path.relative(filePath, from: packageLocation));
      }
    }

    final allLibraryIdentifiers = _classDefinitions.keys.toList();

    // add classes exported in the same library they are defined
    for (final libraryIdentifier in allLibraryIdentifiers) {
      _classExports[libraryIdentifier] = <int>{};
      _classExports[libraryIdentifier]!
          .addAll(_classDefinitions[libraryIdentifier]!);
    }

    var hasFixedPoint = false;
    while (!hasFixedPoint) {
      // initially assume we have a fixed point
      hasFixedPoint = true;
      for (final libraryIdentifier in allLibraryIdentifiers) {
        for (final exportedLibrary in _libraryExports[libraryIdentifier]!) {
          // everything in `exportedLibrary` is also exported in
          // `libraryIdentifier` if it isn't already
          if (!_classExports[libraryIdentifier]!
              .containsAll(_classExports[exportedLibrary]!)) {
            // at least some symbols aren't already exported
            _classExports[libraryIdentifier]!
                .addAll(_classExports[exportedLibrary]!);
            // we made some changes
            hasFixedPoint = false;
          }
        }
      }
    }

    // fill in `_package` with information from _classExports
    for (var library in _package.libraries) {
      library.exportedClasses.addAll(_classExports[library.identifier]!);
    }

    return _package;
}
```

This is the same, just that now everything is private and exists inside a single function...

jonasfj commented on lib/src/package_analysis/shape.dart:
```diff
@@ -0,0 +1,56 @@
+
+part 'shape.g.dart';
+
+@JsonSerializable()
+class PackageShape {
```
Consider adding `@sealed` from `package:meta`.
```suggestion
@sealed
@JsonSerializable()
class PackageShape {
```

This is a nice way to indicate to anyone else that they should not extend or implement you subclasses of your class.
This is a data class, I hope we can all agree that no-one should ever extend it.

CicadaCinema replied:
I suppose the same should go for all the other Shape classes?

jonasfj replied:
Yeah, in fact it probably applies to almost all classes you create... maybe not the AnalysisContext, that one is nice that people can implement themselves, if they want to report warnings differently.

CicadaCinema committed:
clean up classes in shape.dart

CicadaCinema committed:
fix directory structure (in tests/ at least)
also make tests pass given the previous commit

CicadaCinema committed:
opt for nested functions rather than one Summarizer class

CicadaCinema committed:
further correct file names and paths

jonasfj commented on test/package_analysis/testdata/export_cycle.yaml:
```diff
@@ -0,0 +1,45 @@
+      {
+        "uri": "package:test_package/b.dart",
+        "exportedClasses": [
+          2,
+          1
```
Sort maybe?

CicadaCinema commented on lib/src/package_analysis/shape.dart:
```diff
@@ -0,0 +1,76 @@
+
+  /// `Set` of [ClassShape.id] elements, where each one corresponds to a class
+  /// exported in this library.
+  final Set<int> exportedClasses;
```
In my mind, the `exported` prefix tells us that this property uses integer ids as pointers.

CicadaCinema committed:
add functionality for summarising top-level variables, top-level functions and properties
also add some tests

CicadaCinema commented on lib/src/package_analysis/shape.dart:
```diff
@@ -0,0 +1,115 @@
+/// A Shape for describing a class method or a top-level function.
+@sealed
+@JsonSerializable()
+class MethodShape {
```
Maybe this can be called something like `ExecutableShape` seeing as analyzer has the class `ExecutableElement` which is a superclass for both `FunctionElement` and `MethodElement`?

jonasfj replied:
++, or ` CallableShape`, but yes `ExecutableShape` is very nice! :D

CicadaCinema commented on lib/src/package_analysis/summary.dart:
```diff
@@ -0,0 +1,177 @@
+  }
+
+  PropertyShape summarizeProperty(PropertyAccessorElement property) {
+    return PropertyShape(name: property.variable.name);
```
add explanatory comment

CicadaCinema committed:
throw an error if package_config.json doesn't exist

CicadaCinema committed:
implement sorting and re-indexing of classes to obtain a canonical JSON format
testing this feature is kind of difficult...

CicadaCinema committed:
fix analysis error in sort test

jonasfj commented on lib/src/package_analysis/shape.dart:
```diff
@@ -0,0 +1,115 @@
+import 'package:json_annotation/json_annotation.dart';
+import 'package:meta/meta.dart';
+
+part 'shape.g.dart';
```
nit: perhaps we should rename the files `shapes.dart` and `shapes.g.dart`

jonasfj commented on lib/src/package_analysis/shape.dart:
```diff
@@ -0,0 +1,115 @@
+  Map<String, dynamic> toJson() => _$PackageShapeToJson(this);
+}
+
+/// A Shape for describing a library, which may not be contained in a single
+/// file.
```
```suggestion
/// A Shape for describing a library, which may consist of multiple parts.
```

jonasfj commented on lib/src/package_analysis/shape.dart:
```diff
@@ -0,0 +1,115 @@
+
+  /// `Set` of [ClassShape.id] elements, where each one corresponds to a class
+  /// exported in this library.
+  Set<int> exportedClasses;
```
```suggestion
  final Set<int> exportedClasses;
```

jonasfj commented on lib/src/package_analysis/shape.dart:
```diff
@@ -0,0 +1,115 @@
+  /// exported in this library.
+  Set<int> exportedClasses;
+
+  LibraryShape(
+      {required this.uri,
+      required this.getters,
+      required this.setters,
+      required this.functions,
+      required this.exportedClasses});
```
I would personally use trailing comma, but I know others feel differently, do whatever you like :shrug: 

I just wanted to point out that it can also be written as:
```suggestion
  LibraryShape({
    required this.uri,
    required this.getters,
    required this.setters,
    required this.functions,
    required this.exportedClasses,
  });
```

jonasfj commented on lib/src/package_analysis/shape.dart:
```diff
@@ -0,0 +1,115 @@
+  ///
+  /// Classes will be referenced using their [id] in other shapes, such as
+  /// [LibraryShape].
+  int id;
```
```suggestion
  final int id;
```

jonasfj commented on lib/src/package_analysis/summary.dart:
```diff
@@ -0,0 +1,215 @@
+  PackageAnalysisContext packageAnalysisContext,
+  String packageLocation,
+) async {
+  final package =
+      PackageShape(libraries: <LibraryShape>[], classes: <ClassShape>[]);
```
Just saying trailing comma is also allowed here, but do whatever you want :rofl: 
```suggestion
  final package = PackageShape(
    libraries: <LibraryShape>[],
    classes: <ClassShape>[]
  );
```

jonasfj commented on lib/src/package_analysis/summary.dart:
```diff
@@ -0,0 +1,215 @@
+      PackageShape(libraries: <LibraryShape>[], classes: <ClassShape>[]);
+
+  /// Ids of classes defined in a given library
+  final classDefinitions = <String, Set<int>>{};
```
Just a suggestion, it's doesn't matter...
```suggestion
  /// A mapping from [LibraryShape.uri] to [ClassShape.id] of each class defined in a given library.
  final libraryUriToDefinedClassID = <String, Set<int>>{};
```

But for all these maps, it might be preferable to write them as: `<something>To<something>`, like:
 * `libraryUriToExportedLibraryUris` instead of `libraryExports`
 * `libraryUriToExportedClassIDs` instead of `classExports`

-----

Given that all of these are `libraryUriTo....` maybe we could do something like:
```dart

/// Structure that holds intermediate information used during construction of [PackageShape].
///
/// This is only intended for use inside [summarizePackage].
class _LibraryEntry {
  final Set<int> definedClassIDs;
  final List<String> exportedLibraryUris;
  final Set<int> exportedClassIDs;
}

Future<PackageShape> summarizePackage(
  PackageAnalysisContext packageAnalysisContext,
  String packageLocation,
) async {
  ...

  // A map from [Library.uri] to [_LibraryEntry] which holds information about:
  //  * defined classes,
  //  * exported library uris,
  //  * exported classes.
  //
  // This is used for fixed-point computation of transitively exported classes.
  final libraries = <String, _LibraryEntry>{};
```

This is just to make the code easier to read / follow... we could even a step further can rename `_LibraryEntry` to `_LibraryShapeBuilder`, because that's kind of what it is... It's a thing that is used to build a `LibraryShape`.
```dart

class _LibraryShapeBuilder {
  final Set<int> definedClassIDs;
  final List<String> exportedLibraryUris;
  final Set<int> exportedClassIDs;

  ShapeBuilder build() => LibraryShape(....);
}
```

Please feel free to ignore all these suggestions... I'm just seeing patterns :rofl: and we don't have to over-engineer this -- keeping it simple is certainly a good idea, I think this is something we could consider doing later...

jonasfj commented on lib/src/package_analysis/summary.dart:
```diff
@@ -0,0 +1,215 @@
+          .sorted(Comparable.compare)
+          .toSet();
+    }
+  }
```
I propose putting this into a separate library / file, and making it a post-processing step.
```dart
/// Given a [PackageShape] create a new [PackageShape] structure where all [ClassShape.id] values
/// have been canonicalized by sorting by class names blablabla. 
PackageShape canonicalizePackageShape(PackageShape summary) {
  ...
  return PackageShape(...); // create a new PackageShape.
}
```

And maybe put this logic into `canonicalize_shape.dart` or something like that.

Yes, this is a bit wasteful since we are re-creating all the objects.. so we'll use more memory.
But it allows us to keep `ClassShape.id` as `final int id`, and if in the future we decide that this is expensive and we don't need to do it, we can just stop calling the function.

jonasfj commented on lib/src/package_analysis/summary.dart:
```diff
@@ -0,0 +1,215 @@
+
+    classDefinitions[identifier] = classIds;
+    libraryExports[identifier] = libraryElement.exportedLibraries
+        .map((library) => library.identifier)
```
here you need to look at `show` and `hide`

It is possible to do:
```dart
export 'foo.dart' show retry;
```
This exports only the symbol named `retry` from `foo.dart`.

We have to account for this, that's why `libraryExports` is probably too simple...
You need to remember what symbols are show/hide...

jonasfj commented on lib/src/package_analysis/summary.dart:
```diff
@@ -0,0 +1,215 @@
+      // match [packageLocation]/lib/*.dart
+      // but exclude [packageLocation]/lib/src/*.dart
+      if (!(path.isWithin(path.join(packageLocation, 'lib'), filePath) &&
+          !path.isWithin(path.join(packageLocation, 'lib', 'src'), filePath) &&
```
We might need to do something here too..
Let's say we have `lib/foo.dart` which exports `lib/src/foo_impl.dart`, then clearly you need to scan through all of `foo_impl.dart`.

But the `LibraryShape` we get from `lib/src/foo_impl.dart` is not part of the `PackageShape`.

This means that we risk ending up in a situation where `PackageShape.classes` contains classes that aren't exported from anything, but was add there because we had to scan `lib/src/foo_impl.dart` because we didn't know yet if it was exported or not..

We can try to do an analysis upfront to figure out what is exported, before we create the `ClassShape` objects, or we can simple make a post-processing step there filters out any `PackageShape.classes` entries that aren't referenced elsewhere in the `PackageShape`.

CicadaCinema committed:
refactor analysis using `exportNamespace.definedNames`

CicadaCinema committed:
sprinkle in some trailing commas
also fix nits

CicadaCinema committed:
add test for showing/hiding specific symbols when exporting a library
this requires no modification to the analysis logic

CicadaCinema committed:
add test for sorting classes by library name in normalization function
(this comes up only when the names of the two classes being compared are identical)

CicadaCinema committed:
create test for exporting symbols from any libraries in ./src
also improve docstring for `normalizePackageShape`

CicadaCinema committed:
make normalization function more concise by introducing an extension method

CicadaCinema committed:
rename MethodShape to ExecutableShape

jonasfj commented on lib/src/package_analysis/normalize_shape.dart:
```diff
@@ -0,0 +1,67 @@
+            exportedClasses: library.exportedClasses
+                .map((classId) => oldIdToNewId[classId]!)
+                .sorted(Comparable.compare)
+                .toSet(),
```
Should we consider making this a `List` because `Set` is probably unordered..
And it's nice to store things ordered when saving to JSON, maybe?

jonasfj commented on lib/src/package_analysis/summary.dart:
```diff
@@ -0,0 +1,129 @@
+    return PropertyShape(name: property.variable.name);
+  }
+
+  ClassShape summarizeClassElement(ClassElement classElement) {
```
Maybe we should start by checking if `package.classes` contains a `ClassShape.id == classElement.id`?

So we don't create duplicates..

CicadaCinema replied:
yeah I think I agree - rather than relying on Sets it might be better to do this kind of check here

CicadaCinema replied:
putting this logic into `summarizeClassElement` will probably result in making its return type nullable - I've performed this check outside this function:

```dart
package.classes.addAll(classes.where((thisClass) => !package.classes.contains(thisClass)));
```

jonasfj replied:
You're still creating a ClassShape object with the same ID as one that exists in `pacakge.classes` I would suggest not doing that... and just return the value that exists in `package.classes`.

```dart
final shape = package.classes.firstWhere((s) => s.id == classElement.id, else: () => null);
if (shape != null) {
  return shape;
}
```

jonasfj commented on lib/src/package_analysis/shapes.dart:
````diff
@@ -0,0 +1,122 @@
+  /// ```
+  final String uri;
+
+  final List<PropertyShape> getters;
+  final List<PropertyShape> setters;
+
+  /// The top-level functions exported in this library.
+  final List<ExecutableShape> functions;
````
I think we should consider creating `id`s for these as well... because of importing two conflicting definitions work, we probably need to consider it a breaking change to re-defined a previous re-exported symbol.

CicadaCinema committed:
replace Sets with Lists in all Shapes

CicadaCinema committed:
implement ids for top-level getters, setters and functions

jonasfj commented on lib/src/package_analysis/shapes.dart:
```diff
@@ -0,0 +1,168 @@
+}
+
+@internal
+abstract class BasicShape {
```
```suggestion
abstract class GlobalShapeBase {
```

jonasfj replied:
because all global things have IDs, or we could call it something else.. but Base is the typical suffix.

CicadaCinema committed:
rename BasicShape to GlobalShapeBase

CicadaCinema committed:
avoid creating duplicate ClassShape, GlobalPropertyShape and FunctionShape objects
